### PR TITLE
[KIP-396] List offsets implementation

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

**Review by emasab — CHANGES_REQUESTED** [review]
Here are the comments, sent a PR where I've done most of the suggestions for testing them.

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1096]: Why are we not using `TopicPartition` from the input given by the user?

File: src/confluent_kafka/admin/_listoffsets.py
Comment (pranavrth) [inline src/confluent_kafka/admin/_listoffsets.py:19]: Need more understanding on this. We will discuss on a call.

**Review by pranavrth — CHANGES_REQUESTED** [review]
First Round of Review. Checking test cases.

File: examples/adminapi.py
Comment (mahajanadhitya) [reply examples/adminapi.py:692]: Actually it is not but Emanuele has made so the request is more writable by the user. But I do feel like it is a overkill.

File: src/confluent_kafka/admin/__init__.py
Comment (mahajanadhitya) [reply src/confluent_kafka/admin/__init__.py:1065]: Please suggest something in things like this otherwise it remains just stuck I feel this is an apt. description.

File: src/confluent_kafka/admin/__init__.py
Comment (mahajanadhitya) [reply src/confluent_kafka/admin/__init__.py:1096]: Because it might have offset or other value set and we do not want that in our key so only topic and partition should be set and for the internal usecase the offset_spec is stored in the offset.

File: examples/adminapi.py
Comment (emasab) [reply examples/adminapi.py:692]: To try different values of that option. It's the first argument in other examples (LK) too

File: src/confluent_kafka/admin/__init__.py
Comment (emasab) [reply src/confluent_kafka/admin/__init__.py:1071]: It's similar to the javadoc
https://kafka.apache.org/35/javadoc/org/apache/kafka/clients/admin/Admin.html#listOffsets(java.util.Map)

**Review by emasab — APPROVED** [review]
Thanks @mahajanadhitya @pranavrth . This looks good, let's wait for @pranavrth approval or for other comments.

File: examples/adminapi.py
Comment (philipnee) [inline examples/adminapi.py:689]: it can only be 0 here no? len(args) < 1.

File: src/confluent_kafka/admin/__init__.py
Comment (emasab) [reply src/confluent_kafka/admin/__init__.py]: This has been changed because the rich compare method only checks the topic and partition

File: examples/adminapi.py
Comment (pranavrth) [inline examples/adminapi.py:772]: There is no restriction like this in Java. Is empty map allowed there?

File: src/confluent_kafka/_model/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/_model/__init__.py:145]: Move them to field level. It doesn't render properly in docs.

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1193]:
```suggestion
        topic_partition_offsets_list = [_TopicPartition(topic_partition.topic, int(topic_partition.partition), int(offset_spec._value)) for topic_partition, offset_spec in topic_partition_offsets.items()]
```

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1190]: `topic_partition_offsets_list` `instead`.

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1188]: Generally, we should not change the original variable. I would suggest to change the variable to `isolation_level_value` and use it in the `Admin.c` file

File: src/confluent_kafka/admin/_listoffsets.py
Comment (pranavrth) [inline src/confluent_kafka/admin/_listoffsets.py:140]:
```suggestion
        Not available (-1) when querying for the earliest or the latest offsets.
```

?

File: src/confluent_kafka/src/Admin.c
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:2893]:
```suggestion
                             "isolation_level_value",
```

File: src/confluent_kafka/src/Admin.c
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4319]: This function should take `rd_kafka_ListOffsetsResultInfo_t **` as input instead of `rd_kafka_ListOffsets_result_t *`. Conversion from `rd_kafka_ListOffsets_result_t` to `rd_kafka_ListOffsetsResultInfo_t **` should happen in the case block of `Admin_background_event_cb` as per convention that we are following.

File: src/confluent_kafka/src/Admin.c
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4349]: Please check in other places whether we should handle exception case here?

File: src/confluent_kafka/src/Admin.c
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4698]: Same

File: tests/integration/admin/test_list_offsets.py
Comment (pranavrth) [inline tests/integration/admin/test_list_offsets.py:52]: Let's add a case for no Admin options.

File: tests/test_Admin.py
Comment (pranavrth) [inline tests/test_Admin.py:1019]: Add more test cases for all the validations.

**Review by pranavrth — CHANGES_REQUESTED** [review]
Few more comments. 
Add Changelog.

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1186]: Remove extra line break

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1188]: My point was that we should not touch the input variable. We can omit this line.

File: src/confluent_kafka/src/Admin.c
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4359]: Use `Py_XDECREF` instead for raise cases.

when `i=0`, `result=NULL`, this statement will give segfault. Its safer to use XDECREF in general.

**Review by pranavrth — APPROVED** [review]
LGTM!. Minor nits.

File: src/confluent_kafka/admin/__init__.py
Comment (emasab) [reply src/confluent_kafka/admin/__init__.py:1188]: No it gives an error in that case because there's no `isolation_level` parameter

File: src/confluent_kafka/src/Admin.c
Comment (emasab) [reply src/confluent_kafka/src/Admin.c:4359]: result cannot be NULL here, it's initialized before the for loop and there's no goto before

Comment (pranavrth) [reply src/confluent_kafka/src/Admin.c:4359]: Still I prefer XDECREF. But its fine this way as well.

File: src/confluent_kafka/admin/__init__.py
Comment (pranavrth) [reply src/confluent_kafka/admin/__init__.py:1187]: Didn't get you. I am talking about deleting the key.

Comment (emasab) [reply src/confluent_kafka/admin/__init__.py:1187]: Yeah, if it's not deleted the error comes later when calling with kwargs

Comment (pranavrth) [reply src/confluent_kafka/admin/__init__.py:1187]: I don't understand why and where it gives issue. But its fine either way.

**Pull request merged**
